tests/task_scheduler_test.c

```diff
@@ -0,0 +1,168 @@
+
+    uint64_t current_clock_time;
+    ASSERT_SUCCESS(aws_high_res_clock_get_ticks(&current_clock_time), "Failed to get ticks from high-res clock.");
+
+    struct aws_task task2;
+    task2.fn = (aws_task_fn)2;
+    task2.arg = (void *)2;
+
+    //run 250 ms in the future.
+    uint64_t task2_timestamp = current_clock_time + 250000000;
+    ASSERT_SUCCESS(aws_task_scheduler_schedule_future(&scheduler, &task2, task2_timestamp), 
+            "Schedule task in %lluns in the future failed", task2_timestamp);
+
+    struct aws_task task1;
+    task1.fn = (aws_task_fn)1;
+    task1.arg = (void *)1;
+
+    //run now.
+    ASSERT_SUCCESS(aws_task_scheduler_schedule_now(&scheduler, &task1), "Failed to schedule task1");
+
+    struct aws_task task3;
+    task3.fn = (aws_task_fn)3;
+    task3.arg = (void *)3;
+
+    //run 500 ms in the future.
+    uint64_t task3_timestamp = current_clock_time + 500000000;
+    ASSERT_SUCCESS(aws_task_scheduler_schedule_future(&scheduler, &task3, task3_timestamp), 
+            "Schedule task in %lluns in the future failed.", task3_timestamp);
+
+    struct aws_task task_to_run;
+
+    ASSERT_SUCCESS(aws_task_scheduler_next_task(&scheduler, &task_to_run, 0), 
+            "Task pop on a now() task, should return on the first try");
+
+    ASSERT_TRUE(task1.fn == task_to_run.fn, "Popped task should have been task 1.");
+    ASSERT_TRUE(task1.arg == task_to_run.arg, "Popped task arg should have been task 1.");
+
+    uint64_t timestamp = 0;
+    while (aws_task_scheduler_next_task(&scheduler, &task_to_run, &timestamp) != AWS_OP_SUCCESS) {
+        ASSERT_INT_EQUALS(task2_timestamp, timestamp, "Timestamp should for next run should be %llu", 
+                (long long unsigned)task2_timestamp);
+    }
+
+    ASSERT_TRUE(task2.fn == task_to_run.fn, "Popped task should have been task 2.");
+    ASSERT_TRUE(task2.arg == task_to_run.arg, "Popped task arg should have been task 2.");
+
+    while (aws_task_scheduler_next_task(&scheduler, &task_to_run, &timestamp) != AWS_OP_SUCCESS) {
+        ASSERT_INT_EQUALS(task3_timestamp, timestamp, "Timestamp should for next run should be %llu",
+                (long long unsigned)task3_timestamp);
+    }
+
+    ASSERT_TRUE(task3.fn == task_to_run.fn, "Popped task should have been task 3.");
+    ASSERT_TRUE(task3.arg == task_to_run.arg, "Popped task arg should have been task 3.");
+
+    aws_task_scheduler_clean_up(&scheduler);
+    return 0;
+}
+
+static int test_scheduler_pops_task_fashionably_late(struct aws_allocator *alloc, void *ctx) {
+    struct aws_task_scheduler scheduler;
+    aws_task_scheduler_init_default(&scheduler, aws_high_res_clock_get_ticks, alloc);
+
+    uint64_t current_clock_time;
+    ASSERT_SUCCESS(aws_high_res_clock_get_ticks(&current_clock_time), "high res clock ticks failed");
+
+    struct aws_task task;
+    task.fn = (aws_task_fn)0;
+    task.arg = (void *)0;
+
+    //clocks are in nanoseconds. Add 10 milliseconds to give the test time to run.
```

Comment:
You can easily get a 10ms hiccup in your test if your thread is preempted. A mock clock helps avoid this